File: app/src/main/java/ru/radiationx/anilibria/ui/activities/SettingsActivity.java
package ru.radiationx.anilibria.ui.activities;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.MenuItem;

import org.jetbrains.annotations.NotNull;

import io.reactivex.disposables.CompositeDisposable;
import ru.radiationx.anilibria.App;
import ru.radiationx.anilibria.R;
import ru.radiationx.anilibria.extension.ContextKt;
import ru.radiationx.anilibria.model.data.holders.AppThemeHolder;
import ru.radiationx.anilibria.ui.fragments.settings.SettingsFragment;


/**
 * Created by radiationx on 25.12.16.
 */

public class SettingsActivity extends BaseActivity {

    private AppThemeHolder appThemeHolder = App.injections.getAppThemeHolder();
    private CompositeDisposable disposables = new CompositeDisposable();
    private AppThemeHolder.AppTheme currentAppTheme = appThemeHolder.getTheme();

    @NotNull
    public static Intent getIntent(@NotNull Context context) {
        return new Intent(context, SettingsActivity.class);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        currentAppTheme = appThemeHolder.getTheme();
        setTheme(ContextKt.getPrefStyleRes(currentAppTheme));
        setContentView(R.layout.activity_settings);

        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setHomeButtonEnabled(true);
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setDisplayShowTitleEnabled(true);
            actionBar.setTitle("Настройки");
        }

        getSupportFragmentManager().beginTransaction().replace(R.id.fragment_content, new SettingsFragment()).commit();

        disposables.add(
                appThemeHolder
                        .observeTheme()
                        .subscribe(appTheme -> {
                            if (currentAppTheme != appTheme) {
                                currentAppTheme = appTheme;
                                recreate();
                            }
                        })
        );
    }


    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home)
            finish();
        return true;
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        disposables.clear();
    }
}

File: app/src/main/java/ru/radiationx/anilibria/ui/activities/SettingsActivity.kt
package ru.radiationx.anilibria.ui.activities

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import io.reactivex.disposables.CompositeDisposable
import ru.radiationx.anilibria.R
import ru.radiationx.anilibria.di.extensions.injectDependencies
import ru.radiationx.anilibria.extension.getPrefStyleRes
import ru.radiationx.anilibria.model.datasource.holders.AppThemeHolder
import ru.radiationx.anilibria.ui.fragments.settings.SettingsFragment
import javax.inject.Inject


/**
 * Created by radiationx on 25.12.16.
 */

class SettingsActivity : BaseActivity() {

    @Inject
    lateinit var appThemeHolder: AppThemeHolder

    private val disposables = CompositeDisposable()

    private lateinit var currentAppTheme: AppThemeHolder.AppTheme
    override fun onCreate(savedInstanceState: Bundle?) {
        this.injectDependencies()
        currentAppTheme = appThemeHolder.getTheme()
        setTheme(currentAppTheme.getPrefStyleRes())
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_settings)

        val actionBar = supportActionBar
        if (actionBar != null) {
            actionBar.setHomeButtonEnabled(true)
            actionBar.setDisplayHomeAsUpEnabled(true)
            actionBar.setDisplayShowTitleEnabled(true)
            actionBar.title = "Настройки"
        }

        supportFragmentManager.beginTransaction().replace(R.id.fragment_content, SettingsFragment()).commit()

        disposables.add(
                appThemeHolder
                        .observeTheme()
                        .subscribe { appTheme ->
                            if (currentAppTheme !== appTheme) {
                                currentAppTheme = appTheme
                                recreate()
                            }
                        }
        )
    }


    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home)
            finish()
        return true
    }

    override fun onDestroy() {
        super.onDestroy()
        disposables.clear()
    }

    companion object {

        fun getIntent(context: Context): Intent {
            return Intent(context, SettingsActivity::class.java)
        }
    }
}
